File: NoTomorrow/Features/Workout/ProgramBrowserSheet.swift
import SwiftUI
import SwiftData

/// "Browse programs" from the Train tab: the built-in programs (`ProgramLibrary`) with their description, days a week
/// and routine count; tapping one shows its routines and exercises and "Add N routines", which writes them through
/// `RoutineStore.add` (unique names, appended to the routine list).
struct ProgramBrowserSheet: View {
    var programs: [TrainingProgram] = ProgramLibrary.all

    @Environment(\.dismiss) private var dismiss
    @Environment(\.modelContext) private var context
    @State private var selected: TrainingProgram?
    /// Library id → localized exercise name, for the detail's lines.
    @State private var exerciseNames: [String: String] = [:]

    var body: some View {
        VStack(spacing: 0) {
            header
            ScrollView {
                VStack(alignment: .leading, spacing: 0) {
                    if let selected {
                        detail(selected)
                    } else {
                        list
                    }
                }
                .padding(.horizontal, NT.Spacing.screenH)
                .padding(.bottom, NT.Spacing.section)
            }
            .id(selected?.id ?? "")
            if let selected {
                addBar(selected)
            }
        }
        .ntScreenBackground()
        .presentationBackground(NT.Colors.ground)
        .presentationDragIndicator(.visible)
        .onAppear(perform: loadNames)
    }

    // MARK: Header

    /// Back (on a program) · Programs · Done.
    private var header: some View {
        ZStack {
            Text("program.title")
                .font(NT.Fonts.headline).foregroundStyle(NT.Colors.ink).lineLimit(1)
            HStack {
                if selected != nil {
                    Button { selected = nil } label: {
                        HStack(spacing: 4) {
                            Image(systemName: "chevron.left").font(.system(size: 15, weight: .semibold))
                            Text("common.back").font(NT.Fonts.body)
                        }
                        .foregroundStyle(NT.Colors.ink2)
                        .frame(minHeight: NT.Size.control)
                    }
                }
                Spacer()
                Button { dismiss() } label: {
                    Text("common.done").font(NT.Fonts.body).foregroundStyle(NT.Colors.ink2)
                        .frame(minHeight: NT.Size.control)
                }
            }
        }
        .padding(.horizontal, NT.Spacing.screenH)
        .padding(.top, 16)
        .padding(.bottom, 12)
    }

    // MARK: List

    private var list: some View {
        VStack(alignment: .leading, spacing: 10) {
            Text("program.intro")
                .font(NT.Fonts.subheadline).foregroundStyle(NT.Colors.ink2)
                .fixedSize(horizontal: false, vertical: true)
                .padding(.bottom, 6)
            ForEach(programs) { program in
                Button { selected = program } label: { programCard(program) }
                    .buttonStyle(PressScale())
            }
        }
        .padding(.top, 4)
    }

    private func programCard(_ program: TrainingProgram) -> some View {
        NTCard {
            VStack(alignment: .leading, spacing: 6) {
                HStack(alignment: .firstTextBaseline, spacing: 8) {
                    Text(TrainingProgram.text(program.name))
                        .font(NT.Fonts.headline).foregroundStyle(NT.Colors.ink).lineLimit(1)
                    Spacer(minLength: 8)
                    Image(systemName: "chevron.right")
                        .font(.system(size: 13, weight: .semibold)).foregroundStyle(NT.Colors.ink3)
                }
                Text(TrainingProgram.text(program.summary))
                    .font(NT.Fonts.footnote).foregroundStyle(NT.Colors.ink2)
                    .multilineTextAlignment(.leading)
                    .fixedSize(horizontal: false, vertical: true)
                meta(program)
                    .padding(.top, 4)
            }
        }
    }

    /// The level badge, then "3× a week · 3 routines".
    private func meta(_ program: TrainingProgram) -> some View {
        HStack(spacing: 8) {
            Badge(text: LocalizedStringKey(program.level.titleKey))
            Text(verbatim: Self.metaLine(program))
                .font(NT.Fonts.caption).foregroundStyle(NT.Colors.ink2).tabular().lineLimit(1)
        }
    }

    static func metaLine(_ program: TrainingProgram) -> String {
        let perWeek = String(format: String(localized: "program.perWeek"), locale: .current, program.daysPerWeek)
        return perWeek + " · " + WorkoutStrings.routines(program.routines.count)
    }

    // MARK: Detail

    private func detail(_ program: TrainingProgram) -> some View {
        VStack(alignment: .leading, spacing: 0) {
            Text(TrainingProgram.text(program.name))
                .font(NT.Fonts.title1).foregroundStyle(NT.Colors.ink)
                .fixedSize(horizontal: false, vertical: true)
            Text(TrainingProgram.text(program.summary))
                .font(NT.Fonts.subheadline).foregroundStyle(NT.Colors.ink2)
                .fixedSize(horizontal: false, vertical: true)
                .padding(.top, 6)
            meta(program)
                .padding(.top, 10)
            ForEach(program.routines) { day in
                SectionHeader(title: LocalizedStringKey(day.name),
                              trailing: Text(verbatim: WorkoutStrings.exercises(day.items.count)))
                    .padding(.top, NT.Spacing.section)
                    .padding(.bottom, 4)
                VStack(spacing: 0) {
                    ForEach(Array(day.items.enumerated()), id: \.offset) { index, line in
                        if index > 0 { Hairline() }
                        lineRow(line)
                    }
                }
            }
        }
        .padding(.top, 4)
    }

    /// "Bench Press ······ 3 × 8  2:00".
    private func lineRow(_ line: TrainingProgram.Line) -> some View {
        HStack(spacing: 12) {
            Text(exerciseNames[line.exercise] ?? line.exercise.replacingOccurrences(of: "_", with: " "))
                .font(NT.Fonts.subheadline).foregroundStyle(NT.Colors.ink).lineLimit(1)
            Spacer(minLength: 8)
            Text(verbatim: "\(line.sets) × \(line.reps)")
                .font(NT.Fonts.footnote).foregroundStyle(NT.Colors.ink2).tabular()
            Text(verbatim: RoutineRest.label(line.rest))
                .font(NT.Fonts.caption).foregroundStyle(NT.Colors.ink3).tabular()
                .frame(minWidth: 34, alignment: .trailing)
        }
        .padding(.vertical, 10)
    }

    // MARK: Add

    private func addBar(_ program: TrainingProgram) -> some View {
        PrimaryButton(title: LocalizedStringKey(WorkoutStrings.addRoutines(program.routines.count))) {
            RoutineStore.add(program, in: context)
            Haptics.tap()
            dismiss()
        }
        .padding(.horizontal, NT.Spacing.screenH)
        .padding(.vertical, 8)
        .background(NT.Colors.ground)
    }

    private func loadNames() {
        let ids = Array(Set(programs.flatMap(\.exerciseIDs)))
        let found = (try? context.fetch(FetchDescriptor<Exercise>(predicate: #Predicate { ids.contains($0.id) }))) ?? []
        exerciseNames = Dictionary(found.map { ($0.id, $0.localizedName) }, uniquingKeysWith: { first, _ in first })
    }
}

File: android/app/src/main/java/app/notomorrow/feature/workout/ProgramBrowserSheet.kt
package app.notomorrow.feature.workout

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.key
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.produceState
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.hapticfeedback.HapticFeedbackType
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.LocalHapticFeedback
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import app.notomorrow.data.entity.ExerciseEntity
import app.notomorrow.designsystem.Badge
import app.notomorrow.designsystem.Hairline
import app.notomorrow.designsystem.NT
import app.notomorrow.designsystem.NTCard
import app.notomorrow.designsystem.NtIcon
import app.notomorrow.designsystem.NtIcons
import app.notomorrow.designsystem.NtSheet
import app.notomorrow.designsystem.NtText
import app.notomorrow.designsystem.PrimaryButton
import app.notomorrow.designsystem.SectionHeader
import app.notomorrow.designsystem.TabularText
import app.notomorrow.designsystem.ntPlainClickable
import app.notomorrow.designsystem.pressScale
import app.notomorrow.designsystem.sfIconSize
import app.notomorrow.service.localizedName
import app.notomorrow.util.Fmt
import app.notomorrow.util.NtKeys
import app.notomorrow.util.S

/**
 * "Browse programs" from the Train tab — `NoTomorrow/Features/Workout/ProgramBrowserSheet.swift`:
 * the built-in programs ([ProgramLibrary]) with their description, level, days a week and routine
 * count; tapping one shows its routines and exercises and "Add N routines", which writes them
 * through [RoutineStore.addProgram] (unique names, appended to the routine list).
 *
 * @param loadExercises library id → exercise, for the detail's localized names.
 * @param onAdd adds the program; the lambda it gets resolves a routine's catalog key to its name.
 */
@Composable
fun ProgramBrowserSheet(
    loadExercises: suspend (Collection<String>) -> Map<String, ExerciseEntity>,
    onAdd: (TrainingProgram, (String) -> String) -> Unit,
    onDismiss: () -> Unit,
) {
    val context = LocalContext.current
    val haptics = LocalHapticFeedback.current
    val programs by produceState(emptyList<TrainingProgram>()) { value = ProgramLibrary.load(context) }
    val exercises by produceState(emptyMap<String, ExerciseEntity>(), programs) {
        if (programs.isNotEmpty()) value = loadExercises(programs.flatMapTo(mutableSetOf()) { it.exerciseIds })
    }
    var selectedId by rememberSaveable { mutableStateOf<String?>(null) }
    val selected = programs.firstOrNull { it.id == selectedId }

    NtSheet(onDismiss = onDismiss, showsHandle = true, containerColor = NT.Colors.ground) {
        ProgramBrowserHeader(showsBack = selected != null, onBack = { selectedId = null }, onDone = onDismiss)

        // A new scroll position for the list and for each program (`.id(selected?.id)`).
        key(selectedId) {
            Column(
                modifier = Modifier
                    .fillMaxWidth()
                    .weight(1f)
                    .verticalScroll(rememberScrollState())
                    .padding(horizontal = NT.Spacing.screenH)
                    .padding(bottom = NT.Spacing.section),
            ) {
                if (selected == null) {
                    ProgramList(programs = programs, onSelect = { selectedId = it.id })
                } else {
                    ProgramDetail(program = selected, exercises = exercises)
                }
            }
        }

        if (selected != null) {
            val count = selected.routines.size
            PrimaryButton(
                title = stringResource(NtKeys.addRoutines(count), count),
                modifier = Modifier.padding(horizontal = NT.Spacing.screenH).padding(vertical = 8.dp),
                onClick = {
                    onAdd(selected) { key -> ProgramKeys.text(key)?.let(context::getString) ?: key }
                    haptics.performHapticFeedback(HapticFeedbackType.Confirm)
                    onDismiss()
                },
            )
        }
    }
}

/** A catalog key of `programs.json` in the app's language (the key itself if unknown). */
@Composable
private fun programText(key: String): String {
    val id = ProgramKeys.text(key) ?: return key
    return stringResource(id)
}

/** Back (on a program) · Programs · Done. */
@Composable
private fun ProgramBrowserHeader(showsBack: Boolean, onBack: () -> Unit, onDone: () -> Unit) {
    Box(
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = NT.Spacing.screenH)
            .padding(top = 16.dp, bottom = 12.dp),
        contentAlignment = Alignment.Center,
    ) {
        NtText(text = stringResource(S.program_title), style = NT.Fonts.headline, color = NT.Colors.ink, maxLines = 1)
        Row(Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
            if (showsBack) {
                Row(
                    modifier = Modifier.heightIn(min = NT.Size.control).ntPlainClickable(onClick = onBack),
                    horizontalArrangement = Arrangement.spacedBy(4.dp),
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    NtIcon(NtIcons.ChevronLeft, size = sfIconSize(15f), tint = NT.Colors.ink2)
                    NtText(text = stringResource(S.common_back), style = NT.Fonts.body, color = NT.Colors.ink2)
                }
            }
            Spacer(Modifier.weight(1f))
            Box(
                modifier = Modifier.heightIn(min = NT.Size.control).ntPlainClickable(onClick = onDone),
                contentAlignment = Alignment.Center,
            ) {
                NtText(text = stringResource(S.common_done), style = NT.Fonts.body, color = NT.Colors.ink2)
            }
        }
    }
}

/** The intro line, then one card per program. */
@Composable
private fun ProgramList(programs: List<TrainingProgram>, onSelect: (TrainingProgram) -> Unit) {
    Column(Modifier.padding(top = 4.dp), verticalArrangement = Arrangement.spacedBy(10.dp)) {
        NtText(
            text = stringResource(S.program_intro),
            modifier = Modifier.padding(bottom = 6.dp),
            style = NT.Fonts.subheadline,
            color = NT.Colors.ink2,
        )
        programs.forEach { program ->
            key(program.id) {
                NTCard(modifier = Modifier.pressScale(onClick = { onSelect(program) })) {
                    Row(Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
                        NtText(
                            text = programText(program.name),
                            modifier = Modifier.weight(1f),
                            style = NT.Fonts.headline,
                            color = NT.Colors.ink,
                            maxLines = 1,
                        )
                        Spacer(Modifier.width(8.dp))
                        NtIcon(NtIcons.ChevronRight, size = sfIconSize(13f), tint = NT.Colors.ink3)
                    }
                    NtText(
                        text = programText(program.summary),
                        modifier = Modifier.padding(top = 6.dp),
                        style = NT.Fonts.footnote,
                        color = NT.Colors.ink2,
                    )
                    ProgramMeta(program, Modifier.padding(top = 10.dp))
                }
            }
        }
    }
}

/** The level badge, then "3× a week · 3 routines". */
@Composable
private fun ProgramMeta(program: TrainingProgram, modifier: Modifier = Modifier) {
    val count = program.routines.size
    Row(modifier, horizontalArrangement = Arrangement.spacedBy(8.dp), verticalAlignment = Alignment.CenterVertically) {
        Badge(text = programText(program.levelKey))
        TabularText(
            text = stringResource(S.program_perWeek, program.daysPerWeek) + " · " +
                stringResource(NtKeys.routineCount(count), count),
            style = NT.Fonts.caption,
            color = NT.Colors.ink2,
        )
    }
}

/** The program's name, description and meta, then each routine with its exercises. */
@Composable
private fun ProgramDetail(program: TrainingProgram, exercises: Map<String, ExerciseEntity>) {
    Column(Modifier.padding(top = 4.dp)) {
        NtText(text = programText(program.name), style = NT.Fonts.title1, color = NT.Colors.ink)
        NtText(
            text = programText(program.summary),
            modifier = Modifier.padding(top = 6.dp),
            style = NT.Fonts.subheadline,
            color = NT.Colors.ink2,
        )
        ProgramMeta(program, Modifier.padding(top = 10.dp))
        program.routines.forEach { day ->
            SectionHeader(
                title = programText(day.name),
                modifier = Modifier.padding(top = NT.Spacing.section, bottom = 4.dp),
                trailing = stringResource(NtKeys.exerciseCount(day.items.size), day.items.size),
            )
            day.items.forEachIndexed { index, line ->
                if (index > 0) Hairline()
                ProgramLine(line = line, exercise = exercises[line.exercise])
            }
        }
    }
}

/** "Bench Press ······ 3 × 8  2:00". */
@Composable
private fun ProgramLine(line: TrainingProgram.Line, exercise: ExerciseEntity?) {
    Row(
        modifier = Modifier.fillMaxWidth().padding(vertical = 10.dp),
        horizontalArrangement = Arrangement.spacedBy(12.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        NtText(
            text = exercise?.localizedName() ?: line.exercise.replace('_', ' '),
            modifier = Modifier.weight(1f),
            style = NT.Fonts.subheadline,
            color = NT.Colors.ink,
            maxLines = 1,
        )
        TabularText(
            text = "${line.sets} ${Fmt.TIMES} ${line.reps}",
            style = NT.Fonts.footnote,
            color = NT.Colors.ink2,
        )
        TabularText(
            text = if (line.rest <= 0) stringResource(S.routine_restDefault) else Fmt.clock(line.rest),
            modifier = Modifier.widthIn(min = 34.dp),
            style = NT.Fonts.caption.copy(textAlign = TextAlign.End),
            color = NT.Colors.ink3,
        )
    }
}
